com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java

```diff
@@ -45,25 +46,24 @@
 
     private BreakpointManager manager = new BreakpointManager();
 
-    private boolean isHcrInitialized = false;
-
     @Override
     public List<Command> getTargetCommands() {
         return Arrays.asList(Command.SETBREAKPOINTS);
     }
 
     @Override
-    public CompletableFuture<Response> handle(Command command, Arguments arguments, Response response, IDebugAdapterContext context) {
-        // TODO: This part logic should be part of post launching handler.
-        //       Add post launch logic and move the reinstall breakpoints logic there.
-        if (!isHcrInitialized) {
-            IHotCodeReplaceProvider hcrProvider = context.getProvider(IHotCodeReplaceProvider.class);
-            hcrProvider.onClassRedefined((typenames) -> {
-                this.reinstallBreakpoints(context, typenames);
+    public void initialize(IDebugAdapterContext context) {
+        IDebugRequestHandler.super.initialize(context);
+        IHotCodeReplaceProvider provider = context.getProvider(IHotCodeReplaceProvider.class);
+        provider.getEventHub()
+            .filter(event -> event.getEventType() == EventType.END)
+            .subscribe(event -> {
+                reinstallBreakpoints(context, null);
```

Comment:
Need pass the redefined classes for reinstall.

Reply:
Please check the updated version.